x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/TransportGrantAction.java

```diff
@@ -50,11 +59,48 @@ protected void executeWithGrantAuthentication(GrantRequest grantRequest, ActionL
                 );
                 return;
             }
+
+            final String runAsUsername = grantRequest.getGrant().getRunAsUsername();
+
+            final ActionListener<Authentication> authenticationListener = ActionListener.wrap(authentication -> {
+                if (authentication.isRunAs()) {
+                    final String effectiveUsername = authentication.getEffectiveSubject().getUser().principal();
+                    if (runAsUsername != null && false == runAsUsername.equals(effectiveUsername)) {
+                        // runAs is ignored
+                        listener.onFailure(
+                            new ElasticsearchStatusException("the provided grant credentials do not support run-as", RestStatus.BAD_REQUEST)
+                        );
+                    } else {
+                        // Authentication can be run-as even when runAsUsername is null.
+                        // This can happen when the authentication itself is a run-as client-credentials token.
+                        assert runAsUsername != null || "access_token".equals(grantRequest.getGrant().getType());
+                        authorizationService.authorize(
+                            authentication,
+                            AuthenticateAction.NAME,
+                            new AuthenticateRequest(effectiveUsername),
+                            ActionListener.wrap(ignore2 -> listener.onResponse(authentication), listener::onFailure)
+                        );
+                    }
+                } else {
+                    if (runAsUsername != null) {
+                        // runAs is ignored
+                        listener.onFailure(
+                            new ElasticsearchStatusException("the provided grant credentials do not support run-as", RestStatus.BAD_REQUEST)
+                        );
+                    } else {
+                        listener.onResponse(authentication);
+                    }
+                }
+            }, listener::onFailure);
+
+            if (runAsUsername != null) {
+                threadContext.putHeader(AuthenticationServiceField.RUN_AS_USER_HEADER, runAsUsername);
+            }
             authenticationService.authenticate(
                 actionName,
                 grantRequest,
```

Comment:
I'd like to change this `grantRequest` to an `AuthenticateRequest` similar to how it is done in the authZ code. But it is not possible currently because `AuthenticateRequest` requires username and it may not be known if the provided credential is an accessToken. I think the right thing to do is to refactor `AuthenticateRequest` to drop the `username` field. We no longer support transport client. It is not possible for end users to set this field to a different value. `AuthenticateRequest` should always work for the current authentication object and this logic should just be enforced regardless of the username. I will raise another PR to make this change if you agree with the proposal.

Reply:
The change to drop username from `AuthenticateRequest` seems reasonable.
I'm not sure that I see the significance of using Authenticate here though. We could, but it doesn't seem to be strictly better.